File: clients/desktop/src/main/version-manager.ts
/**
 * 版本管理器
 * 
 * 对等 Android VersionManager
 * 负责检查远程版本、读取和保存本地版本
 */

import * as fs from 'fs';
import * as path from 'path';
import * as https from 'https';
import { app } from 'electron';
import { AppConfig } from './config';

export class VersionManager {
  private versionFile: string;

  constructor() {
    // 版本号存储在 userData 目录
    this.versionFile = path.join(app.getPath('userData'), 'version.txt');
  }

  /**
   * 从远程获取版本号
   */
  async fetchRemoteVersion(): Promise<string> {
    return new Promise((resolve, reject) => {
      https.get(AppConfig.versionUrl, (res) => {
        if (res.statusCode !== 200) {
          reject(new Error(`Failed to fetch version: ${res.statusCode}`));
          return;
        }

        let data = '';
        res.on('data', (chunk) => {
          data += chunk;
        });

        res.on('end', () => {
          try {
            const json = JSON.parse(data);
            resolve(json.version);
          } catch (err) {
            reject(new Error(`Invalid version JSON: ${data}`));
          }
        });
      }).on('error', (err) => {
        reject(err);
      });
    });
  }

  /**
   * 读取本地版本号
   */
  readLocalVersion(): string | null {
    if (fs.existsSync(this.versionFile)) {
      return fs.readFileSync(this.versionFile, 'utf-8').trim();
    }
    return null;
  }

  /**
   * 保存本地版本号
   */
  saveLocalVersion(version: string): void {
    fs.writeFileSync(this.versionFile, version, 'utf-8');
    console.log(`[VersionManager] Saved version: ${version}`);
  }
}

File: clients/app/android/app/src/main/java/com/weiqi/app/VersionManager.kt
package com.weiqi.app

import android.content.Context
import com.weiqi.app.util.Logger
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONObject
import java.io.File

/**
 * 版本管理器
 * 负责检查远程版本、读取和保存本地版本
 */
class VersionManager(
    private val context: Context,
    private val client: OkHttpClient
) {
    
    companion object {
        private const val TAG = "VersionManager"
        private val VERSION_URL: String get() = AppConfig.versionUrl
    }

    /**
     * 从远程获取版本号
     * 使用 use 自动关闭 Response
     */
    fun fetchRemoteVersion(): String {
        val request = Request.Builder()
            .url(VERSION_URL)
            .build()

        // ✅ 使用 use 自动关闭 Response
        return client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                throw Exception("Failed to fetch version: ${response.code}")
            }
            val body = response.body?.string() ?: "{}"
            val json = JSONObject(body)
            json.getString("version")
        }
    }

    /**
     * 读取本地版本号
     */
    fun readLocalVersion(): String? {
        val versionFile = File(context.filesDir, "version.txt")
        return if (versionFile.exists()) {
            versionFile.readText().trim()
        } else {
            null
        }
    }

    /**
     * 保存本地版本号
     */
    fun saveLocalVersion(version: String) {
        val versionFile = File(context.filesDir, "version.txt")
        versionFile.writeText(version)
        Logger.d(TAG, "Saved version: $version")
    }
}
